apis/network/v1beta1/dnsmasq_types.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// AnnotationHostnameKey -
+	AnnotationHostnameKey = "dnsmasq.network.openstack.org/hostname"
+
+	// DNSDataLabelSelector - label selector to identify config maps with hosts data
+	DNSDataLabelSelector = "dnsmasqhosts"
+)
+
+// DNSMasqOption defines allowed options for dnsmasq
+type DNSMasqOption struct {
+	// +kubebuilder:validation:Enum=server;rev-server;srv-host;txt-record;ptr-record;rebind-domain-ok;naptr-record;cname;host-record;caa-record;dns-rr;auth-zone;synth-domain
+	Key    string   `json:"key"`
+	Values []string `json:"values"`
+}
+
+// DNSMasqSpec defines the desired state of DNSMasq
+type DNSMasqSpec struct {
+	// +kubebuilder:validation:Optional
+	// DNSMasq Container Image URL
+	ContainerImage string `json:"containerImage"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1
+	// Replicas - DNSMasq Replicas
+	Replicas int32 `json:"replicas"`
+
+	// +kubebuilder:validation:Optional
+	// Debug - enable debug for different deploy stages. If an init container is used, it runs and the
+	// actual action pod gets started with sleep infinity
+	Debug DNSMasqDebug `json:"debug,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Options allows to customize the dnsmasq instance
+	Options []DNSMasqOption `json:"options,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// NodeSelector to target subset of worker nodes running this service. Setting
+	// NodeSelector here acts as a default value and can be overridden by service
+	// specific NodeSelector Settings.
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// ExternalEndpoints, expose a VIP using a pre-created IPAddressPool
+	ExternalEndpoints []MetalLBConfig `json:"externalEndpoints,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="dnsdata"
```

Comment:
Should this default be the same as the value set on L31?

Reply:
no, L31 is the key of the label. this is the value. its probably confusing that in L28 I call it AnnotationHostnameKey, but in L31 don't name it key. I'll change that to be consistent

Reply:
done